[language] new recursive descent move-lang parser

bob-wilson opened the pull request:
## Motivation

This replaces the lalrpop parser for move-lang in the same way as for the IR compiler. This includes the equivalent of the latest IR compiler lexer, changes to the binary expression operator precedence (to match Rust), and simplifications of the ParseError type. It also updates ParseError::User to do something reasonable, but further improvements to parsing errors are clearly going to be needed.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

This passes the existing move-lang tests. More parsing tests should be added, but I'd like to get this merged first.

tnowacki commented on language/move-lang/src/parser/syntax.rs:
```diff
@@ -1,4 +1,1412 @@
+) -> Result<(), ParseError<usize, &'static str>> {
+    if tokens.peek() != tok {
+        return Err(ParseError::InvalidToken {
+            location: tokens.start_loc(),
```
could we just add a really really simple improvement to the errors? 
```
let peeked = tokens.peek();
if tok != peeked {
  return Err(ParserError::InvalidToken { location: tokens.start_loc(), expected: tok, actual: peeked })
}
```

bob-wilson replied:
I added back ParserError::UnrecognizedToken  with something like that.

tnowacki commented on language/move-lang/src/parser/syntax.rs:
```diff
@@ -1,4 +1,1412 @@
+//
+// This works for either Tok::NameValue or Tok::MNameValue
+// tokens. The caller is responsible for checking the token kind.
+
```
nit, why the space between function and the comment?

bob-wilson replied:
I was trying to make the comments with the grammar stand out, and then figured it would be good to be consistent. Would you prefer to avoid those blank lines?

tnowacki replied:
I prefer large section headers (https://github.com/libra/libra/blob/master/language/move-lang/src/typing/ast.rs#L19) 

And then comments on each function (no space)
It might also be worthwhile using `///` for rustdocs

bob-wilson replied:
OK, well, my goal was consistency, so I should try to match your style. I'll kill the blank lines and see if there are large sections that merit headers. There's only one public function in that file (parse_file_string) so I can add rustdoc comments for that one.

bob-wilson replied:
I made those changes. What do you think, @tnowacki ?

tnowacki commented on language/move-lang/src/parser/syntax.rs:
```diff
@@ -1,4 +1,1412 @@
+        }
+        _ => {
+            return Err(ParseError::InvalidToken {
+                location: tokens.start_loc(),
```
I see for some of these, having the expected token is hard.... but I'm kind of hesitant to let it go in without something here. 

I want to get rid of lalrpop, but its not really blocking me to have it right now. 

Maybe we could do tests and better parser messages as the next step (i.e. soon). Thoughts?

bors-libra commented:
:umbrella: The latest upstream changes (presumably 6ffb7a04383270b534bd09fa39e4cd2602d7567a) made this pull request unmergeable. Please resolve the merge conflicts.

bob-wilson commented:
I updated this for the latest changes to syntax.lalrpop. I also fixed up the parser errors to be closer to what lalrpop provided. The clever "token_match" macro trick that @tnowacki provided helps, with the slight downside that I had to replicate some match patterns so that each one handles a single kind of token. I randomly broke some tests to look at the error messages, and the results confirm my suspicion that listing the expected tokens is not very good in practice. I'll go with this for now, but just as one example, where it doesn't work well, consider this function definition: when parsing a comma separated list of function parameters, after the Nth parameter, the code is currently written to check for the closing parenthesis

tnowacki commented on language/move-lang/src/parser/syntax.rs:
```diff
@@ -1,4 +1,1396 @@
+        },
+        Tok::U64Value => {
+            let v = parse_value(tokens)?;
+            Exp_::Value(v)
```
@tzakian Any macro ideas on how to combine these cases?

bors-libra commented:
:umbrella: The latest upstream changes (presumably 9331cf51d0aa8753added6f1b0e4e5e136e0c4bc) made this pull request unmergeable. Please resolve the merge conflicts.

huitseeker commented on language/move-lang/src/lib.rs:
```diff
@@ -160,55 +160,43 @@ fn check_errors(errors: Errors) -> Result<(), Errors> {
-        )
+        let mut msg = String::new();
+        for (i, e) in expected.iter().enumerate() {
+            let sep = match i {
```
I appreciate the art in "or", but would like to point to the existence of [join](https://doc.rust-lang.org/std/primitive.slice.html#method.join)

bob-wilson replied:
Yes, if we don't care about using a different separate between the last elements ("or"), it would make sense to use join here. I'm not a big fan of this particular error message regardless, but I was trying to improve the readability as much as I could.

bob-wilson replied:
I changed to use join for now.

huitseeker commented on language/move-lang/src/parser/lexer.rs:
```diff
@@ -0,0 +1,375 @@
+
+// Return the length of the substring matching [a-zA-Z0-9_]. Note that
+// this does not do any special check for whether the first character
+// starts with a number, so the caller is responsible for any additional
```
OK, how do we make this [quirk](https://doc.rust-lang.org/reference/identifiers.html) more visible? perhaps make this a doc comment (`///`) so it prints in doc?

bob-wilson replied:
The function is private within this file and rust-doc is only for public functions, right?

huitseeker replied:
huh, there seems to be an option for it, but it may not be worth buying into:
https://github.com/rust-lang/cargo/pull/5543

bob-wilson commented:
@huitseeker makes a good suggestion about a smoke benchmark but I don't think that is specific to parsing. It applies to all phases of the compiler, and in fact, parsing is relatively simple compared to things like type inference and borrow checking, so if we're going to test compile time, it seems like we should just test the whole thing end-to-end. That would definitely be a good thing to do at some point.

huitseeker commented:
@bob-wilson if you want non-regression testing, [criterion](https://github.com/bheisler/criterion.rs) is your firend!

tnowacki commented:
I don't think the smoke/perf benchmark is our main priority right now (i.e. we have way more pressing matters). We really don't care about perf at all right now (outside of slowing down the tests for the repo).

bors-libra commented:
:umbrella: The latest upstream changes (presumably 49788a86d8c3d4f9b8f83e796625e086adcc02d6) made this pull request unmergeable. Please resolve the merge conflicts.

bob-wilson commented:
I've updated this to work with the latest syntax and also fixed a few issues exposed by my pending PRs for parser tests. I'm still not happy with the current style of error messages: the "expected" tokens often reflect only a subset of the valid input options because the parsing decisions are made in stages and the errors are only reported at the last point. E.G., an expression term may be followed by a binary operator and another term, but if it is not, then the parser concludes that it has reached the end of the expression and goes looking for whatever is expected next -- the possibility of having a binary operator is not reflected in the set of "expected" tokens. Providing an accurate set of "expected" tokens is not actually very helpful, so this won't be a real problem when we can add more helpful diagnostics. (I'm eager to get to that next.)

bors-libra commented:
:umbrella: The latest upstream changes (presumably 98fba3198670a55c2403f9349867ffbdaed9a0a2) made this pull request unmergeable. Please resolve the merge conflicts.

bob-wilson commented:
@bors-libra r=tnowacki

bors-libra commented:
:pushpin: Commit da9e70e has been approved by `tnowacki`

<!-- @bors-libra r=tnowacki da9e70e56e164951311924a1abed69449ad271e0 -->

bors-libra commented:
:hourglass: Testing commit da9e70e56e164951311924a1abed69449ad271e0 with merge 384839c2fe202f68f4c6a64ba0a44ec6479de067...

bors-libra commented:
:sunny: Test successful - [checks-circle_commit_workflow](https://circleci.com/workflow-run/2c402ae9-b091-4791-a661-b2aa04870e61?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-checks-link)
Approved by: tnowacki
Pushing 384839c2fe202f68f4c6a64ba0a44ec6479de067 to master...
